File: src/main/java/com/smartdialer/provider/TelecomProvider.java
package com.smartdialer.provider;

import com.smartdialer.domain.ProviderHealth;
import com.smartdialer.domain.ProviderEventType;
import java.util.UUID;
import java.util.List;

public interface TelecomProvider {
    ProviderCallResult initiateCall(UUID callId, String borrowerId);
    boolean cancelCall(UUID callId);
    ProviderHealth getHealth();
    default List<com.smartdialer.domain.ProviderEvent> pollEvents(UUID callId) { return List.of(); }
    record ProviderCallResult(boolean accepted, String providerCallId, ProviderEventType initialEvent, String failureReason) { }
}

File: src/provider/TelecomProvider.ts
// ============================================================================
// SmartDialer — Telecom Provider Interface
// ============================================================================
// The dialer must NOT know provider-specific internals.
// All provider implementations satisfy this interface.
// ============================================================================

export interface CallRequest {
  callId: string;
  phoneNumber: string;
  campaignId: string;
}

export interface InitiateCallResponse {
  providerCallId: string;
  status: 'initiated' | 'failed';
  latencyMs: number;
}

export interface ProviderEvent {
  eventId: string;
  providerCallId: string;
  eventType: 'RINGING' | 'ANSWERED' | 'CONNECTED' | 'COMPLETED' | 'FAILED' | 'CANCELLED';
  sequenceNumber: number;
  timestamp: string;
  payload: Record<string, unknown>;
}

export interface TelecomProvider {
  /** Unique provider name */
  readonly name: string;

  /** Initiate an outbound call. May throw on failure/timeout. */
  initiateCall(request: CallRequest): InitiateCallResponse;

  /** Cancel a previously initiated call. */
  cancelCall(providerCallId: string): void;

  /** Get current provider health assessment. */
  getHealthStatus(): 'HEALTHY' | 'DEGRADED' | 'UNHEALTHY';
}
